### ISSUE TITLE
Version 0.5 causes switch/case symbols to show up

### ISSUE THREAD

**Issue by roblabla** [boricj/ghidra-delinker-extension#8]
Hey, thanks for the great extension!

I'm currently working on using this along with [objdiff](https://github.com/encounter/objdiff) in my decompilation project [th06](https://github.com/happyhavoc/th06) to enable a better diffing workflow.

I have something working great with delinker version 0.4. However, in version 0.5, the delinker now contains a bunch of extra symbols that add a ton of noise, and don't seem to be of great value. In particular

Delinker 0.4:
<img width="581" alt="image" src="https://github.com/user-attachments/assets/48a02303-0f77-43e2-a86e-75c3009d1009">

Delinker 0.5:
<img width="598" alt="image" src="https://github.com/user-attachments/assets/1f8c7e23-4ead-4ea4-a01b-5079201ccd65">

Those symbols seem to be inserted automatically by ghidra during switch/case analysis.

Would it make sense perhaps to detect them and not include them in the symbol table of the generated object file?

**Comment by boricj** [comment]
Hi,

Given all the bugs I've fixed over the past three months (see #6 for a partial history), the COFF exporter wasn't fit for purpose until version 0.5.0 ; I would not trust any results coming from version 0.4.0.

This is not a use-case I have foreseen and I'm not sure it can be fully accommodated. This extension was developed with the idea that it's not only possible to delink executables, but that the result could also be interchangeable with the original object file *as far as the linker is concerned*. This means that the contents of the exported object file can differ, because there are multiple equivalent ways to encode its information.

For example, comparing `ascii-table.o` from the [ascii-table Linux MIPS ELF](https://github.com/boricj/ghidra-delinker-extension/tree/f47c58c4f62b3718ad9e0c4886aba450662f793e/src/test/resources/ascii-table/reference/elf_linux_freestanding/mipsel) test case against its exported equivalent with objdiff show several discrepancies across multiple functions. This is because the compiler and the exporter chose different ways to refer to the same offsets within a section, in a manner that resulted in different values for the addends of the relocations ; specifically, the compiler used unnamed section-relative relocations whereas the exporter used traditional named symbols.

Link-time optimizations and relaxations are another source of possible discrepancies. One particularly tricky example on MIPS hoists instructions inside branch delay slots, [requiring some migraine-inducing workarounds](https://github.com/boricj/ghidra-delinker-extension/blob/f47c58c4f62b3718ad9e0c4886aba450662f793e/src/main/java/ghidra/app/analyzers/relocations/MipsCodeRelocationSynthesizer.java#L477) if those instructions happened to have HI16 relocations. I haven't seen anything remotely as drastic on x86, but I'm vaguely aware of peephole optimizations that can eliminate some indirections through the PLT or GOT at link time by rewriting instructions and relocation types.

Going back to your initial issue, if the `Give dynamic symbols external visibility` option was left unchecked in the exporter these dynamic symbols should be local and not global. Alternatively, the exporter could be modified to not emit local symbols and generate section-relative relocations instead, something that I'm willing to add as an option. That might be enough for your particular use-case, but I'm doubtful that perfect delinking in general can be achieved with my extension, given its policy of not using any manual, delinker-specific annotations.

**Comment by roblabla** [comment]
> This is because the compiler and the exporter chose different ways to refer to the same offsets within a section, in a manner that resulted in different values for the addends of the relocations ; specifically, the compiler used unnamed section-relative relocations whereas the exporter used traditional named symbols.

This is fine. Objdiff can be configured to ignore relocation differences. This allows it to paper over slight differences in relocation encoding that ultimately result in the same value (at the expense of potentially not seeing actual differences "hidden" by a relocation, but that's a tradeoff we can afford to take). And anyways the diffing doesn't need to be "perfect" - we're using it as a tool to guide our re-implementation towards having the correct assembly. We're currently using `diff` on assembly dumps, which is much noiser due to not having the exact same memory layout (something we'll work on at a later phase of the decompilation).

> Going back to your initial issue, if the Give dynamic symbols external visibility option was left unchecked in the exporter these dynamic symbols should be local and not global. 

Unfortunately, I tried both with and without "Give dynamic symbols external visibility", and it does not get rid of the "switchD" symbols. When it's enabled, I additionally get symbols for `LAB_` symbols, so the checkmark does do _something_, but it doesn't get rid of those symbols. Those are probably not considered dynamic - they actually show up in the "Symbol Tree" after all, unlike `LAB_` labels.

**Comment by roblabla** [comment]
On a similar note, labels that were renamed to give a more useful name also show up as symbols in the delinker. I feel like those should definitely not show up - just as goto labels don't show up as symbols in a normal COFF file. They also did not show up in 0.4.

<img width="507" alt="image" src="https://github.com/user-attachments/assets/bd497891-5019-46ca-aab8-f331490bca43">

<img width="669" alt="image" src="https://github.com/user-attachments/assets/aeec920d-06c8-434b-88e0-bda2af635034">

I assume this is the same root problem.

**Comment by roblabla** [comment]
Rather than a regex-based solution, could we perhaps have an option to only export symbols in the executable section if they have a matching Function? This should get rid of the goto labels/switchcase, while leaving the symbols for all the functions and the data.

It's obviously not a perfect heuristic (some programs might store data in the .text section/segment), but I feel like it should work decently enough for our use-cases.

**Comment by boricj** [comment]
Hmm... We can check if a function contains the address of a symbol. If so, then the symbol can be made local unless it's at the entrypoint. That should get rid of all the labels within functions and I don't see a downside that would cause problems later on.

As for the switch data labels, the problem remains. I can't use your suggested heuristic because of the sheer scope of this extension. It supports two object file formats, two ISAs and to my knowledge was tested with artifacts from three different platforms (Linux, Windows, PlayStation) spanning 25 years. That's a _very_ delicate balancing act to pull off and it will only get worse as time goes on. From experience, taking shortcuts inside this code base **never** ends well, it's unusually exacting.

That's why I'm looking for solutions which not only work for one use-case, but are correct for all use-cases. Unless there's a better proposal, an option for a regex that matches local symbols (by default set to `switchD_.+::switchdataD_.+`) is the least bad idea I can think of for symbols that aren't inside a function.

**Comment by boricj** [comment]
I've pushed some commits on master to add options that should give these symbols static visibility. Can you check if this catches all the symbols you want gone? Next step will be to make them section-relative and trim them from the symbol table, but it'll require some refactorings first to make it happen.

**Comment by roblabla** [comment]
Yes, it worked perfectly! Only the more "relevant" symbols show up now.

<img width="567" alt="image" src="https://github.com/user-attachments/assets/0688eedf-b376-4106-b4ac-9969c935af69">

**Comment by boricj** [comment]
Fixed by:
* 6ce0489bb0398acbb76226b22ef670282b03a4d4 (symbols within a function)
* 648b4e0d3593bd54d4521f5687a90e0fd7e77490 (symbols matching a regex)
